### ISSUE TITLE
2.5.1 update logic and the use of self-referenced queries as input for value annotations

### ISSUE THREAD

**Issue by gesinn-it** [SemanticMediaWiki/SemanticMediaWiki#2549]
### Setup and configuration

- SMW version: 2.5.1
- MW version: 1.27.x
- PHP version: 7
- DB (MySQL etc.): MySQL

### Issue

I create pages with Page Forms writing a template. As part of the template, I set a property depending on the result of an `#ask` query:

```
{{#vardefine:partOfSite|{{#ask:[[-Part Of Building::{{FULLPAGENAME}}]]|?Part Of Site=|mainlabel=-|link=none|format=list}} }} 
{{#set:Part Of Site={{#var:partOfSite}}}}
```

I'm aware of [0] which is circumnavigated using [1]. But this should not be relevant here.

Starting with 2.5.1, property `Part Of Site` is not created anymore. I tried with downgrading to `2.5.0` and `2.4.5`. Both versions work as expected.

I tried `$smwgEnabledQueryDependencyLinksStore = true;` because I thought that this is related. No change.

Any idea, what change could have caused this issue?

[0] https://www.semantic-mediawiki.org/wiki/Help:Self-Referencing
[1] https://www.mediawiki.org/wiki/Extension:SemanticDependencyUpdater

**Comment by kghbln** [comment]
Just to make sure: This also does not work in 2.5.3?

**Comment by gesinn-it** [comment]
@kghbln not working with 2.5.3

**Comment by mwjames** [comment]
> @kghbln not working with 2.5.3

Standard unit and integration tests pass for #set, the involvement of
#var .. or any other extension is not tested and also not part of SMW
core, and unless there is a stack trace or bisect there is not much
assistance we can provide.

PS: I have on several occasions here (and on wiki) explained that
relying on self-references values might not result in the expected
result because it depends on when and how values are available by the
time #ask is executed. Since we rely on deferred updates that process
can even more ambiguous due to different execution paths run by
MediaWiki.

PSS: There are some technical "tricks" you might be able to use but I
have no plans to look at them more closely in near future. For
example, you could track a #ask that contains a self-reference and
schedule an update job (a simple null edit isn't enough).

**Comment by gesinn-it** [comment]
... before using `#var` I was directly using the result of `#ask` in `#set`. Same effect, not working since 2.5.1. For me, this pattern has been working very stable until 2.5.1 and in many setups. I'm wondering what has caused this breaking now...

**Comment by mwjames** [comment]
> PS: I have on several occasions here (and on wiki) explained that

#1696, #2375, #1478

**Comment by gesinn-it** [comment]
@mwjames thanks for pointing me to some more details. I'll try a few things and drop my findings here...

**Comment by gesinn-it** [comment]
`$smwgEnabledDeferredUpdate = false;` did not have any effect

**Comment by gesinn-it** [comment]
I found, that the changed behavior was introduced with f5a30ddca2a9976f5d8504a9f4e51256e9fe7a6f

**Comment by gesinn-it** [comment]
to be more precise:

```diff
+		if ( $this->skipUpdateOn( $latestRevID ) ) {
+			return $applicationFactory->getMediaWikiLogger()->info( __METHOD__ . " (Found rev:$latestRevID, skip update)" );
+		}
```

**Comment by mwjames** [comment]
This change was necessary due to changes in LinksUpdate where it is
transformed in  `EnqueueableDataUpdate` and would force updates to be
scheduled twice depending on when it is run by the `DeferredUpdates`
(or if is run as Job). Especially changes to `DeferredUpdates` (in MW)
made it necessary for us the ensure we only run on an unknown revision
and for all others we skip the process to avoid duplicate updates.

1.27 is a bit more ordered in terms when updates are scheduled
sequential but 1.28 (or 1.29) use an opportunistic plan so there is no
guarantee on when an update is run.

**Comment by gesinn-it** [comment]
> avoid duplicate updates

@mwjames do you see any chance, to make the behavior configurable? In my case, if the duplicate updates would just have performance issues without breaking anything else, I would prefer this.

**Comment by mwjames** [comment]
> It would be cool, if we could keep both, the old behavior and the changes made in this commit.

On what condition?

Furthermore, it shows that your use case may work in 1.27 but there is no guarantee that it will work with 1.29 or later. We are concerned with update consistency in SMW core and changes in MW required us to pay more attention on the sequence of updates so that we run our own transaction as soon as possible when `LinksUpdate` is executed (which itself is a deferred update) but avoid a possible issue when the `LinskUpdate` is planned as `refreshLinksPrioritized`.

Self-referenced use cases require values to be stored first then at some point to be re-run with an update (e.g. by `RefreshLinksJob`) again to pick-up the data and set the value. This may work or not in some releases (by relying on some update sequence that is not defined or better not defined as dependency of each other) but without a formalized approach those scenario easily break since you rely on a hidden assumption that those update sequences remain stable which is clearly not the case.

By formalized approach I mean, a process that is defined and it is independent of some assumptions made about updates to be run twice or in sequence.

> I was directly using the result of #ask in #set

There is a reason why we don't test (per integration) things like #ask and #set (see below) because we not assume that updates run more than once per revision (as outlined before, updates would need to be run at least twice for the stated example, first to store the value of `TestAnno` and the second time to allow #ask/#show to query the value and set the value to `TestAnno2`, `TestAnno3`).

```
[[TestAnno::Foo]]

[[TestAnno2::{{#show: {{FULLPAGENAME}} |?TestAnno |mainlabel=- |link=none}}]]

{{#set:
 |TestAnno3={{#show: {{FULLPAGENAME}} |?TestAnno |mainlabel=- |link=none}}
}}
```

PS: MediaWiki does change and you deal with issues like https://phabricator.wikimedia.org/T153618 or https://phabricator.wikimedia.org/T154425 that weren't an issue before hence SMW has to adapt as well.

**Comment by mwjames** [comment]
> do you see any chance, to make the behavior configurable?

It is possible but as outlined above there is no guarantee that it will result in the expected outcome in later releases because you make an assumption about a sequence of events the SMW project does not make. 

What happens when `refreshLinksPrioritized` is no longer run, or modified in different ways so that it doesn't execute `LinksUpdate` or its hooks anymore? 

```
[runJobs] refreshLinksPrioritized TestAnno/2 rootJobTimestamp=20170715161728 useRecursiveLinksUpdate=1 triggeringUser={"userId":0,"userName":"127.0.0.1"} triggeringRevisionId=29367 requestId=17d0e9ff4f16f587f8b16e4b (id=13726,timestamp=20170715161728) STARTING
Title::getRestrictionTypes: applicable restrictions to [[TestAnno/2]] are {edit,move}
Parser cache options found.
ParserOutput cache found.
[smw] [mw.db.queryengine] connection provider with {"read":-1,"write":-2}
[smw] SMW\SQLStore\QueryDependency\QueryDependencyLinksStore::doUpdateDependenciesFrom suppressed (skewed time) for SID 192952
[smw] SMW\SQLStore\QueryDependency\QueryDependencyLinksStore::doUpdateDependenciesFrom suppressed (skewed time) for SID 192952
Title::getRestrictionTypes: applicable restrictions to [[TestAnno/2]] are {edit,move}
Title::getRestrictionTypes: applicable restrictions to [[TestAnno/2]] are {edit,move}
[smw] SMW\ParserData::updateStore (Found rev:29367, skip update)
[DBQuery] RefreshLinksJob::runForTitle: committing on behalf of RefreshLinksJob::run.
[DBQuery] LinksUpdate::incrTableUpdate: committing on behalf of RefreshLinksJob::run.
[DBQuery] LinksUpdate::incrTableUpdate: committing on behalf of RefreshLinksJob::run.
BacklinkCache::queryLinks: got results from DB
[objectcache] Duplicate get(): "mw-master:pcache:idoptions:19977" fetched 2 times
[objectcache] Duplicate get(): "mw-master:pcache:idhash:19977-0!canonical" fetched 2 times
[runJobs] refreshLinksPrioritized TestAnno/2 rootJobTimestamp=20170715161728 useRecursiveLinksUpdate=1 triggeringUser={"userId":0,"userName":"127.0.0.1"} triggeringRevisionId=29367 requestId=17d0e9ff4f16f587f8b16e4b (
```

**Comment by gesinn-it** [comment]
@mwjames I completely agree, that the assumptions I made with a sequence of updates is/was more or less working "by accident". It would definitely be better to have a solution that guarantees order of execution / takes care of dependencies.

Let me summarize some of my thoughts: I see two cases:

1) A new page is **created**, {{FULLPAGENAME}} does not exist during `#ask` or only after MW has created FULLPAGENAME. Here we have a dependency, that the page names needs to be created early. Not sure, if MW takes care about that. `#ask` is run inside an `#set` as nested parser functions. I think the parser takes care of the dependencies in such a case.

2) We **edit** an existing page. In this case, we should not have an issue with FULLPAGENAME.

For the first case, a null edit was sufficient but this seems to be not working anymore. I remember that there has been some notes about null edit in some other posts and in [1]...

Anyway, I will review my data design and double check, if I can get rid of all cases of "self referencing", e.g. with the awesome _Property chains and paths_ feature [0]. Usually, I had injected properties, like the one mentioned in my example, when things are part of a hierarchy, and users want to filter. E.g. a hierarchy of "Location Objects":

- Site
-- Building
--- Room Level
---- Room

On each level in the hierarchy, I want to be able to filter (e.g. with SRF Filtered or Drilldown) for "Part Of Site", thus "Part Of Site" is derived from the parent object. SDU extensions takes care of updating children, otherwise this will fall into pieces when changing a "Part Of" dependency...

[0] https://www.semantic-mediawiki.org/wiki/Help:Property_chains_and_paths
[1] https://www.semantic-mediawiki.org/wiki/Null_edit

**Comment by mwjames** [comment]
> I completely agree, that the assumptions I made with a sequence of updates is/was more or less working "by accident".

#2572 introduces an approach to avoid the "by accident" state therefore closing this issue in light of the provided information.
